### DISCUSSION TITLE
What about using Hugo Pipes generating assets instead of prebuilt?

### DISCUSSION THREAD

**Discussion by razonyang** [razonyang/hugo-theme-bootstrap#391] [Ideas]
At the beginning of the theme design, I thought that prebuilt assets would be friendly for users, but as some issues come up, it may not better than Hugo pipes. Such as custom FontAwesome icons, and it may be impossible to custom Bootstrap components, utilities(CSS classes), since the prebuilt assets purged unused CSS and JS.

**Pros**

- Much more flexible, such as custom Bootstrap components, utilities, FontAwesome icons.
- Smaller bundle size, for example: currently `custom.css` appends style to the bundle instead of overriding.

**Cons**

- Requires Hugo extended version.
- Requires other tools: such as `npm`.

It'll include some breaking changes, such as `custom.css` and `custom.js` will be removed. I'd like to know your opinions before making a decision. @yshalsager, @SimpleAnecdote, @AJ2O, @AidaIssayeva, @iambenzo, @benjivesterby, @chrisrummel, @sagarbehere, @Natenom.

You can also vote this idea via thumb up or down emoji.

**Comment by benjivesterby** [comment]
Would there be an increase in performance of the site? Personally I build mine with extended anyways.

**Reply by razonyang** [reply]
I guess yes, smaller bundle size means performance improvement. Let's take `palettes` parameter as an example, the unexpected palettes will be excluded from the final bundle via Hugo pipes. Currently, the prebuilt bundle always contains all palettes style.

**Comment by yshalsager** [comment]
Sounds okay, I use extended anyways, and use font awesome in custom js as well.

**Comment by sagarbehere** [comment]
@razonyang This sounds okay.

- I use Hugo extended already
- I have made many small modifications to the CSS via `custom.css`. Those modifications are important to my use case. Will there be a way to retain those modifications or apply them in some other way (does not need to be via the `custom.css` file)?

**Reply by razonyang** [reply]
Yep, I'll provide other way to do this.

**Comment by Natenom** [comment]
I am a user, not a developer. So I can't really write much to this.

What I liked about your theme when I found it:
1. No external resources (JS, Fonts) are loaded, which meets my privacy settings.
2. No external tools are used.

As long as 1. does not change or will be configurable I am fine with the changes. :)

2. Depends on the documentation :)

**Reply by razonyang** [reply]
Hugo pipes require other tools for building assets(JS, CSS) bundle at runtime. Such as `npm`, `npm` will download dependencies(JS, CSS) for Hugo pipes, but there aren't extra resources will be included in the final assets bundle. I'm not sure if that does it break your point 1?

**Reply by Natenom** [reply]
But thats only for rendering the website, isn't it? The rendered Website will not refer to resources not located on the website. Thats fine then.

**Comment by razonyang** [comment]
I created a PR #405 for this, it's almost done.
